### ISSUE TITLE
`finalDisplayedUrl` being set to `:`

### ISSUE THREAD

**Issue by vipulnaik** [GoogleChrome/lighthouse#15073]
## FAQ

- [X] Yes, my issue is not about [variability](https://github.com/GoogleChrome/lighthouse/blob/main/docs/variability.md) or [throttling](https://github.com/GoogleChrome/lighthouse/blob/main/docs/throttling.md).
- [X] Yes, my issue is not about a specific accessibility audit (file with [axe-core](https://github.com/dequelabs/axe-core) instead).

## URL

https://www.si.com/
https://www.elle.com/
https://www.newyorker.com/

## What happened?

For a few sites with ads, running the Lighthouse report in non-incognito Google Chrome yields empty reports.

### Example of www.si.com

The Lighthouse section of devtools while the Lighthouse report is running:

<img width="1432" alt="Screenshot 2023-05-11 at 11 51 56 PM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/6ec40bae-5dfa-4770-9292-3a75d193ac7c">

What the Lighthouse section of devtools looks like right after the run finishes -- note the blank page, and the fact that there's nothing after the `-` following the time for the title in the dropdown on the top left (after the time of run) left where the url should be present.

<img width="1434" alt="Screenshot 2023-05-11 at 11 52 14 PM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/03b57104-72d1-4143-af4f-01249a222795">

### Example of www.elle.com

The Lighthouse section of devtools while the Lighthouse report is running:

<img width="1434" alt="Screenshot 2023-05-11 at 11 50 44 PM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/47557554-bceb-4a8c-b1d8-462bfc1006bb">

What the Lighthouse section of devtools looks like right after the run finishes -- note the blank page, and the fact that there's nothing after the `-` following the time  for the title in the dropdown on the top left (after the time of run) where the url should be present.

<img width="1435" alt="Screenshot 2023-05-11 at 11 51 01 PM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/6b1a7407-da1f-44f5-95ea-544398ba1595">

### Example of www.newyorker.com

What the Lighthouse section of devtools looks like right after the run finishes -- note the blank page, and the fact that there's nothing after the `-` following the time  for the title in the dropdown on the top left (after the time of run) where the url should be present.

<img width="1435" alt="Screenshot 2023-05-11 at 11 50 25 PM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/ba9d8c29-78c6-44e9-8f9b-95c0b8b97779">

## What did you expect?

I expected behavior similar to what I see in incognito.

The Lighthouse section of devtools while the Lighthouse report is running:

<img width="1434" alt="Screenshot 2023-05-11 at 11 53 33 PM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/5ef2006c-ac75-4008-a3f7-5199ea72528b">

What the Lighthouse section of devtools looks like right after the run finishes -- note the presence of an actual report on the page, and also note that the title in the dropdown on the top left includes the url:

<img width="1436" alt="Screenshot 2023-05-11 at 11 53 57 PM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/803399d0-8031-41b4-9d98-745fe194baec">

## What have you tried?

I used devtools of devtools to access the browser console panel as well as the sources panel for the Lighthouse run itself.

I noticed this error consistently in the console  (in devtools of devtools)every time a blank report was generated, with an occurrence right at the end of the report generation (so right before the blank report was served up to me):

<img width="1438" alt="Screenshot 2023-05-12 at 12 01 12 AM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/fe2ff758-891e-4f1a-a183-a0075b73d269">

Clicking through took me to sources (in devtools of devtools), and I located the line in report.js. I added a debugger breakpoint and reran the Lighthouse report to catch the error when it happened.

<img width="1433" alt="Screenshot 2023-05-12 at 12 03 07 AM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/2e8026fb-4db7-43a3-a1a9-32ade4b7806a">

Note the value of `e` is `:`, and this is causing `createOrReturnURL` to give an error because `:` is not a url. See also the stack trace summary printed on the right. Working backward through the trace, I got to this line:

<img width="1433" alt="Screenshot 2023-05-12 at 12 03 22 AM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/ee7c30ae-f6c2-4fe5-857e-e6bbab37f3db">

So it looks like the return value of `getFinalDisplayedUrl` is `:`, and that's what gets passed down as `e`. To figure out why `getFinalDisplayedUrl` is returning `:`, I set a debugger breakpoint in that function. Then I reran the Lighthouse report.

Here's what I found. See the `Watch` on the right that includes the full value of `e`, and note that at the point of first entry into this function, the `finalDisplayedUrl` attribute is already `:`. So `getFinalDisplayedUrl` is just reading that value and returning it.

<img width="1437" alt="Screenshot 2023-05-12 at 12 12 19 AM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/e0bb1950-b9a3-4c1b-b2a5-edf6378ff9d4">

So what's responsible for the setting of `finalDisplayedUrl` to `:`? I could not figure this part out conclusively, because searching through sources in devtools of devtools did not find the place it was being set. However, looking at the Lighthouse source code on GitHub suggests that this is likely happening with the line https://github.com/GoogleChrome/lighthouse/blob/49aa1425b83928f08a961e48ebd88a8ffa9da2cf/core/gather/navigation-runner.js#L244

I inferred this because my Lighthouse run was of type navigation, and this seems like the only match in the code where that value is being set and that I expect the code to go through. Unfortunately, this code doesn't seem to be accessible in devtools of devtools.

So unwinding, we get that, in non-incognito Google Chrome Devtools Lighthouse, for some sites with ads:

1. finalDisplayedUrl is being set to `:` somehow; this is likely a result of https://github.com/GoogleChrome/lighthouse/blob/49aa1425b83928f08a961e48ebd88a8ffa9da2cf/core/gather/navigation-runner.js#L244 somehow returning `:` but I could not verify this directly.
2. Therefore, `getFinalDisplayedUrl` in https://github.com/GoogleChrome/lighthouse/blob/main/shared/util.js (specifically, https://github.com/GoogleChrome/lighthouse/blob/49aa1425b83928f08a961e48ebd88a8ffa9da2cf/shared/util.js#L58-L62) returns `:`.
3. Therefore, `createOrReturnURL` in https://github.com/GoogleChrome/lighthouse/blob/main/shared/util.js (specifically, https://github.com/GoogleChrome/lighthouse/blob/49aa1425b83928f08a961e48ebd88a8ffa9da2cf/shared/util.js#L281-L287), gets passed the value `:` and gives an error because `:` is not a url.
4. This causes an uncaught error and seems to be responsible for generating a blank report.

NOTE: See comments for further diagnosis, particularly https://github.com/GoogleChrome/lighthouse/issues/15073#issuecomment-1546202462 suggesting that this is related to the privacy sandbox setting.

### How were you running Lighthouse?

Chrome DevTools (non-incognito)

I haven't had an issue with any of these sites when using incognito Chrome Devtools. I also haven't had an issue using the Node Lighthouse, which is similar to incognito in the sense of having a relatively clean slate.

I don't know what is different between my regular Chrome and incognito; one possibility is a bunch of application state including cookies, local storage, IndexedDB etc. However, I haven't visited www.newyorker.com for a while before today when I ran the Lighthouse reports, and I still got the error on www.newyorker.com in my regular Chrome.

I've also seen an increased rate of Lighthouse errors when making requests to pagespeed.web.dev and/or the PageSpeed Insights API for sites with ads, and I suspect that those failures may be for a related reason. In those cases, the information I get is just that "something went wrong" or `ERRORED_DOCUMENT_REQUEST` or something like that, so it's difficult to map that back to the root cause. PageSpeed Insights API seems to be running Lighthouse 10.2.0, so if the root cause is the same, that suggests it is not yet fixed in 10.2.0.

### Lighthouse Version

10.0.1

I manually looked through the changes since 10.0.1 and none of them seem to address the portions of code that I identified. However, since I can't reproduce the issue with Node Lighthouse, it's a little tricky to test the issue against a newer version of Lighthouse.

`finalDisplayedUrl` and `getFinalDisplayedUrl` seem to have been introduced as part of Lighthouse 10, so this particular issue should not have happened with Lighthouse 9. Indeed, I haven't encountered blank reports occurring with Lighthouse 9.

### Chrome Version

113.0.5672.92

This is the latest Chrome version, released about two days ago.

I also have Lighthouse installed using `npm` and I run it with Node v18.15.0, npm 9.5.0. But I haven't replicated the issue in Node Lighthouse, as noted above.

### OS

Mac Silicon, M1 Air, Mac OS 13.3.1 (a)

**Comment by vipulnaik** [comment]
One suggested fix is to update `getFinalDisplayedUrl` to also validate `finalDisplayedUrl` (to check if it is of type URL, the same way that `createOrReturnURL` runs a check) and to fall back to `finalUrl` if `finalDisplayedUrl` is not of type URL. I am not sure this should be done, though. Perhaps it is better to enforce the validation further upstream? Or perhaps enforcing it here as well is a sound idea?

**Comment by vipulnaik** [comment]
necessarily-wide-alpaca

**Comment by vipulnaik** [comment]
"necessarily-wide-alpaca"

**Comment by connorjclark** [comment]
@paulirish i'm kind of over this bot.

**Comment by connorjclark** [comment]
I cannot repro. This not happening in incognito for you is typically a sign of an extension being the cause. Could you try toggling some extensions in your main Chrome profile and see which one it could be?

**Comment by vipulnaik** [comment]
This is still happening to me after I deactivated all extensions, then quit and restart Chrome.

Chrome extensions that I had, all deactivated (even previously, none of the extensions were *active* on those specific pages because I've configured my Chrome extensions to only be active on click or on specific sites):

<img width="1424" alt="Screenshot 2023-05-12 at 11 29 19 AM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/161be959-e177-469f-bd46-356ab0ac8c78">

The report is blank again:

<img width="1434" alt="Screenshot 2023-05-12 at 11 29 55 AM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/55014383-af95-4ff3-8855-afeec3fa66d2">

I see the same error again in devtools console of devtools:

<img width="1435" alt="Screenshot 2023-05-12 at 11 30 11 AM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/9974e0a1-56b9-473c-bc98-57673f90c4e6">

I can try to triangulate exactly what aspect of my local state (cookies, IndexedDB, etc.) is responsible for this. But regardless of the specific cause, what do you think about falling back to `finalUrl` when `finalDisplayedUrl` exists but isn't a url, as I suggested at https://github.com/GoogleChrome/lighthouse/issues/15073#issuecomment-1545342110? At minimum that would get us past the situation of a completely blank report.

Stay tuned -- I'll post another comment after triangulating the aspect of my user state that is causing this issue.

**Comment by vipulnaik** [comment]
ok here's a fixed version of the previous comment after I disabled *all* extensions (I had left one out). Still seeing the issue:

<img width="1421" alt="Screenshot 2023-05-12 at 11 39 51 AM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/3b87fe41-2f0b-4e85-9b2b-04e23ad83bc9">
<img width="1437" alt="Screenshot 2023-05-12 at 11 40 30 AM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/99eb4883-795f-4795-8caf-43ffc7587c80">
<img width="1436" alt="Screenshot 2023-05-12 at 11 40 40 AM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/e1216bc5-263e-458c-b96a-7d7854418da2">

**Comment by vipulnaik** [comment]
@connorjclark it seems that it isn't related to extensions or application state, but rather to the privacy sandbox setting. I noticed something suspicious with some user sync requests being pending on one page load, and nothing else was pending. This led me to remember that I have the privacy sandbox origin trial enabled. This setting can be toggled at chrome://settings/privacySandbox

<img width="1423" alt="Screenshot 2023-05-12 at 12 35 35 PM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/2a19c273-efc1-4b62-950c-6f20ea1b7de5">

Toggling this on and off, I was able to replicate Lighthouse report failing and succeeding. This time I tried another site https://www.thestreet.com/

Here's the alternating on/off fail/success pattern, demonstrated through a screenshot (see the dropdown on the top left):

<img width="1440" alt="Screenshot 2023-05-12 at 12 39 06 PM" src="https://github.com/GoogleChrome/lighthouse/assets/7758551/0dd490ce-80fd-4eaa-a490-6d446f7c2df4">

**Can you reproduce the issue after toggling privacy sandbox on at chrome://settings/privacySandbox on any of the urls (www.si.com, www.elle.com, www.newyorker.com, www.thestreet.com)?** If you can't reproduce, I'll continue to look into differences in my environment that may be driving this. If you can reproduce, I think you're all set? But let me know if you need anything else.

**Comment by vipulnaik** [comment]
@connorjclark FWIW I'm still seeing this issue in Chrome 114 with Lighthouse 10.1.1: the `finalDisplayedUrl` is being set to `-` when the privacy sandbox setting at chrome://settings/privacySandbox is toggled to on, and therefore the Lighthouse report isn't showing up by the mechanism I outlined above. I think this fix https://github.com/GoogleChrome/lighthouse/issues/15073#issuecomment-1545342110 would still work for such situations though it doesn't get to the root cause of why `finalDisplayedUrl` is being set this way.

**Comment by adamraine** [comment]
I'm seeing this happen consistently on https://espn.com in DevTools now. Will investigate further.

**Comment by adamraine** [comment]
Update on investigation, this appears to be coming from a [fenced frame](https://developer.chrome.com/en/docs/privacy-sandbox/fenced-frame/) (notice the target subtype). Still not sure exactly how this event gets created though. Will file upstream.

```json
{
  "method": "Target.attachedToTarget",
  "params": {
    "sessionId": "221795A5659470D7AC6FCBF3F311D882",
    "targetInfo": {
      "targetId": "F5B7345C9BF9413F71A9B51D22304A82",
      "type": "iframe",
      "title": "",
      "url": "",
      "attached": true,
      "canAccessOpener": false,
      "browserContextId": "6310EBC13B0F02CB468081EC10C1A4B0",
      "subtype": "fenced"
    },
    "waitingForDebugger": true
  },
  "sessionId": "B5668ED04C68E8A4178A04D5EE1EB6E0"
}
```

Frame tree:

```json
{
  "frame": {
    "id": "F5B7345C9BF9413F71A9B51D22304A82",
    "loaderId": "4BDF72DDE41DDD533EFDE652EB563B30",
    "url": ":",
    "domainAndRegistry": "",
    "securityOrigin": "://",
    "mimeType": "text/html",
    "adFrameStatus": {
      "adFrameType": "none"
    },
    "secureContextType": "InsecureScheme",
    "crossOriginIsolatedContextType": "NotIsolatedFeatureDisabled",
    "gatedAPIFeatures": []
  }
}
```

**Comment by adamraine** [comment]
https://bugs.chromium.org/p/chromium/issues/detail?id=1464949

**Comment by adamraine** [comment]
The bandaid fix has landed in the latest Puppeteer and it was merged into DevTools. The root cause is tracked by the chromium issue.

No further action should be necessary in the Lighthouse repo.
